### ISSUE TITLE
About PyTorch Tutorial

### ISSUE THREAD

**Issue by ching9752** [Kuanch/office_hour#3]
我在讀Pytorch的Tutorial中遇到了幾個問題～
出來冒個泡泡問問看助教>___<

resnet50(pretrained=False)中pretrained=False是代表什麼意思
第一篇有寫True是直接引入原本的 ImageNet pretrained model，那pretrained=False是不遷移學習就從零開始嗎？

!wget 不能用
不知道我的虛擬環境被我怎麼了XD 總之就是!會被報錯，我網路查了一下改成這樣（所以這個其實不用解決沒關係XDD)
import wget
url = 'https://raw.githubusercontent.com/EliSchwartz/imagenet-sample-images/master/n02085936_Maltese_dog.JPEG'
file_name = wget.download(url)
print(file_name)

get clone失敗
我查了要用terminal後丟了git clone 之後就卡在黑黑的畫面，不知道該怎麼辦了OvO
(所以我現在都慢慢一個cell複製）

當我打開Jupiter notebook後的終端機頁面就沒有(ching) C:\Users\MAGIC>，我就迷失在黑底白字中了，我要如何回到正常畫面C:\Users\MAGIC>（因為想要pip install)
我現在的做法是不管他，全部關掉重開XDDDDDDDDDDDD

引入狗狗圖片的時候 img = cv2.cvtColor(img2, cv2.COLOR_BGR2RGB)，為什麼你知道狗狗的圖片現在是BGR（我都不知道QQ)然後為什麼順序是轉成RGB，這是默認的規定嗎？

img = img / 127.5
img = img - 1 中的127.5是怎麼來的，以及img = img - 1 是什麼意思QQ

好了我暫時就讀到這個cell就卡住惹，偷看了一下下面覺得不太懂XDDDDDDD
這週會努力繼續推進度的！！
助教上班加油！！！！！

**Comment by Kuanch** [comment]
1. 沒錯，`pretrained=False` 是不遷移學習就從零開始，隨機初始化網絡權重    
    
2. 指令是這樣子的嗎 `git clone https://github.com/Kuanch/office_hour.git`
記得要注意你的資料夾內沒有已經下載的 `office_hour` 哦

補充
每個 repo 都可以這樣下載，url 在右上角綠色的 `Code` 選 HTTPS
![image](https://user-images.githubusercontent.com/42059373/166628921-2653964e-95f6-40ac-80ec-4efb676b7ba6.png)

然後如果我這邊有更新，你可以直接在 `office_hour` 這個資料夾內下 `git pull` 就會拿到最新的東西了
  
3. opencv (cv2) 幾乎都是預設 BGR 圖片，`cv2.imread` 也是

4. 因為 opencv 讀出來的圖片值域是 (0, 255)，我把它轉換到 (0, 1) ，這是因為 Pytorch ImageNet pretrained model 是在這個值域訓練的；算是一種 normalization，對模型訓練穩定度會比較好。

**Comment by ching9752** [comment]
我又乖乖地啃了幾天tutorial感覺有幾個地方看不懂：

1. from torchvision.io import read_image #為什麼不能直接寫 import torchvision
我知道這行特別專指從torchvision.io import某個部分（read_image)，但為什麼不能import整個工具？是因為這樣會導致程式很龐大嗎？還是如果不特別指定引入的部分的話，會沒有辦法執行程式？

2. dataset中，這句："dataset[idx] 時，實際上是 dataset.__getitem__(idx)"這句真的完全看不懂，只有他是這樣嗎？其他人都不會用到index嗎？有底線的方法（例如：__XX__）都是特別的預設方法對不對？不是我們自定義的
其他部分似懂非懂XDDDDDD

4. img_dir的dir是什麼的縮寫啊～～～

5. 為什麼在jupyter notebook中
Img = Image.open()
img
下面可以直接讀出圖片
img = cv2.imread()
Img
下面讀出來是矩陣？Python讀圖片不是全部都是以矩陣的形式嗎，怎麼會有人直接是圖片OAO!!

6. dataloader我有看沒有懂，我決定再多看幾次再提問XDDD

**Comment by Kuanch** [comment]
1. import 一個函式或是整個進來並不會影響記憶體或是存取速度，單純是習慣，我的邏輯是，如果整個引入，那可能性很多，因為全部都可以用，但只引入少數幾個有點像是在告訴人說：「只有這幾個會用到」，所以是可讀性考量。

2. 在 python 當中，只要能夠用 `....[idx]` 這種的，譬如說 `list`，是因為它們都實作 `__getitem__` 這個函式。只要用 `....[idx]` 實際上就是呼叫 `__getitem__` ，再換句話說，你用 `a.__getitem__(idx)` 完全等於 `a[idx]`，反之也是。

    沒錯，有底線的都是 python 預設的方法，有些很奇妙，有用到再去查就好。
3. dir 是 directory 的縮寫
4. 的確本質上都是矩陣，但 `PIL.Image.open` 回傳的是一個物件 `PIL.Image`，jupyter notebook 認識它，只要是它就可以直接顯示出來，但 `cv2.imread()` 就真的只是`numpy.array`矩陣了。順帶一提，`PIL.Image.fromarray` 可以把 numpy 矩陣變成 `PIL.Image`
5. 沒問題
